### PR TITLE
[NEW] Channel avatar

### LINKED ISSUE #18262: [Bug] Hide system messages shows dropdown list only after second turn on
### Description:

**Hide system messages** option in chat settings shows dropdown list only after second turn on.
I saw that behavior on all previous versions since that option appeared in chat settings.

### Steps to reproduce:

1. Go to private chat settings
2. Enable **Hide System Messages** option
3. Disable **Hide System Messages** option
4. Enable **Hide System Messages** option again to show dropdown list

### Expected behavior:

Dropdown list should appear immediately after **Hide system messages** is turned on

### Actual behavior:

hide-sysmes

### Server Setup Information:

- Version of Rocket.Chat Server: 3.4.2
- Operating System: CentOS7
- Deployment Method: docker
- Number of Running Instances: 12
- DB Replicaset Oplog: Enabled
- NodeJS Version: v12.16.1
- MongoDB Version: 4.0.16

### Client Setup Information

- Desktop App or Browser Version: 2.17.9
- Operating System: Windows 10

### PR DESCRIPTION
**Pull request by ggazzo**
## Issue(s)

closes #18262

## Screenshots
image

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Hotfix (a major bugfix that has to be merged asap)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by MartinSchoeler** [comment]
I can upload other file formats and break the server if i upload a .pdf this happens: 
Screen Shot 2020-07-30 at 8 33 16 PM

and I get this error on the server console

```
W20200730-20:33:10.278(-3)? (STDERR) === UnHandledPromiseRejection ===
W20200730-20:33:10.279(-3)? (STDERR) [Error: Input file contains unsupported image format]
W20200730-20:33:10.279(-3)? (STDERR) ---------------------------------
W20200730-20:33:10.279(-3)? (STDERR) Errors like this can cause oplog processing errors.
W20200730-20:33:10.279(-3)? (STDERR) Setting EXIT_UNHANDLEDPROMISEREJECTION will cause the process to exit allowing your service to automatically restart the process
W20200730-20:33:10.279(-3)? (STDERR) Future node.js versions will automatically exit the process
W20200730-20:33:10.279(-3)? (STDERR) =================================

```

**Comment by MartinSchoeler** [comment]
Also if this is done not from the admin view the saving fails and the browser gives this error: 

```
Exception in setTimeout callback: TypeError: Cannot read property 'etag' of undefined
I20200730-20:44:02.133(-3)?     at app/lib/server/functions/setRoomAvatar.js:66:45
I20200730-20:44:02.139(-3)?     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
I20200730-20:44:02.140(-3)?     at packages/meteor.js:550:25
I20200730-20:44:02.140(-3)?     at runWithEnvironment (packages/meteor.js:1286:24)

```

no server crashes this time

File: app/channel-settings/server/methods/saveRoomSettings.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
+/* eslint-disable complexity */
```
Comment (sampaiodiego) [inline app/channel-settings/server/methods/saveRoomSettings.js:1]: 🤔

File: app/channel-settings/client/startup/messageTypes.js [hunk 2]
```diff
@@ -29,6 +29,18 @@ Meteor.startup(function() {
 		},
 	});
 
+	MessageTypes.registerType({
+		id: 'room_changed_avatar',
```
Comment (sampaiodiego) [inline app/channel-settings/client/startup/messageTypes.js:33]: should we start using semantically correct types or keep old schema? 🤔 
```suggestion
		id: 'room_avatar_changed',
```

File: app/channel-settings/client/startup/messageTypes.js [hunk 3]
```diff
@@ -29,6 +29,18 @@ Meteor.startup(function() {
...
+		system: true,
+		message: 'room_changed_avatar',
```
Comment (sampaiodiego) [inline app/channel-settings/client/startup/messageTypes.js:35]:
```suggestion
		message: 'room_avatar_changed',
```

File: server/startup/migrations/v203.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import { Avatars } from '../../../app/models/server';
+
+Migrations.add({
+	version: 203,
+	up() {
+		Avatars.tryDropIndex({ name: 1 });
+		Avatars.tryEnsureIndex({ name: 1 }, { sparse: true });
+
+		Avatars.tryDropIndex({ rid: 1 });
+		Avatars.tryEnsureIndex({ rid: 1 }, { sparse: true });
```
Comment (sampaiodiego) [inline server/startup/migrations/v203.js:11]: `rid` index is new, doesn't need a migration 😬 
```suggestion
```

**Comment by sampaiodiego** [comment]
thanks for pointing it out @Exordian .. but the issue was indeed closed by this specific commit (ecd306c91c5b0d42d44a003c0c6aea1a772a3240) included in this PR, although they seem unrelated 👍

**Comment by inimatic** [comment]
In Directory search the avatars not shown.

**Comment by Saidernot** [comment]
> Also if this is done not from the admin view the saving fails and the browser gives this error:
> 
> ```
> Exception in setTimeout callback: TypeError: Cannot read property 'etag' of undefined
> I20200730-20:44:02.133(-3)?     at app/lib/server/functions/setRoomAvatar.js:66:45
> I20200730-20:44:02.139(-3)?     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
> I20200730-20:44:02.140(-3)?     at packages/meteor.js:550:25
> I20200730-20:44:02.140(-3)?     at runWithEnvironment (packages/meteor.js:1286:24)
> ```
> 
> no server crashes this time

I install 3.6.0 and have this problem.

**Comment by evilrussian** [comment]
> I install 3.6.0 and have this problem.
me too

**Comment by BjoernLudwigPTB** [comment]
> Also if this is done not from the admin view the saving fails and the browser gives this error:

In the logs of 3.7.1 the same error appears and as well only, when non-admins are trying to upload an avatar. For admins the feature works as expected.

**Comment by sampaiodiego** [comment]
@BjoernLudwigPTB you can test the release candidate for 3.8.0 .. https://github.com/RocketChat/Rocket.Chat/releases/tag/3.8.0-rc.5

**Comment by BjoernLudwigPTB** [comment]
> @BjoernLudwigPTB you can test the release candidate for 3.8.0 .. https://github.com/RocketChat/Rocket.Chat/releases/tag/3.8.0-rc.5

Unfortunately I myself cannot. In my organisation I am only administrator on the application side, the deployment is done by colleagues. I asked them to have a look and will report.

**Comment by didacroyo** [comment]
It would be great to be able to set the custom Avatar via API, similarly to  
/api/v1/users.setAvatar

image

**Comment by sampaiodiego** [comment]
In theory you should be able to do that using the `rooms.saveRoomSettings` endpoint (which is not documented).. but I agree we having a `rooms.setAvatar` endpoint makes sense.. we're open for contributions =)

**Comment by didacroyo** [comment]
Hi Diego, thanks for your comment. You mean using an undocumented API REST endpoint or you mean the Realtime API? could you share an example call setting an avatarUrl to a rid? thanks

**Comment by sampaiodiego** [comment]
@didacroyo you can do either.. there is an endpoint at `/api/v1/rooms.saveRoomSettings` that accepts the same payload as the Realtime Method called `saveRoomSettings`.. you can check the source code at https://github.com/RocketChat/Rocket.Chat/blob/74597046809979e7f690631bdc169f148a3351b4/app/channel-settings/server/methods/saveRoomSettings.js#L223

you can send a payload with `roomAvatar` field which should work the same as using the `users.setAvatar` endpoint :wink:

**Comment by jmessy919** [comment]
Just tried adding a channel icon to three different channels on the server. Only 1 time it worked. I am using the latest snap on Ubuntu 20. Not sure why it doesn't work, as my manual installation of another RocketChat server (latest version) never has an issue updating the channel icons.

**Comment by ArturoRejon** [comment]
Hi, @sampaiodiego 

We are using our API to change the avatar of the rooms; but despite of all our efforts, it returns 

`{
    "rid": "REAL_ROOM_ID",
    "success": true
}`

The response is 200 ok; but the avatar in the room isn't changed, even if we restart the service or reload without cache. We are sending this using PHP

```
$curl = curl_init();

curl_setopt_array($curl, array(
  CURLOPT_URL => 'https://OUR_CHAT/api/v1/rooms.saveRoomSettings',
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => '',
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 0,
  CURLOPT_FOLLOWLOCATION => true,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => 'POST',
  CURLOPT_POSTFIELDS =>'{
"rid": "REAL_ROOM_ID",
"roomAvatar":"VALID_IMAGE_URL"
}',
  CURLOPT_HTTPHEADER => array(
    'X-User-Id: REAL_ID_API',
    'X-Auth-Token: REAL_TOKEN',
    'Content-Type: application/json'
  ),
));

$response = curl_exec($curl);

curl_close($curl);
echo $response;

```
And we know that it somehow works, because we can send the key "roomName" and it successfully changes the name; but the image is never shown!!!

Thank you in advance for your help. Regards

**Comment by sampaiodiego** [comment]
hi @ArturoRejon , I'm sorry the delay..

looks like the endpoint accepts a `base64` for room avatar only, despite my previous comments and the documentation..

so instead of `VALID_IMAGE_URL` you'll need to parse the image content and provide a `base64` payload, something like:

```
roomAvatar: "data:image/png;base64,iVBORw0KGg ...... "
```

**Comment by ArturoRejon** [comment]
@sampaiodiego thank you, it works!!

**Pull request merged**